File: php/src/Domain/Topology/ValueObject/ZoneTypeCode.php
<?php

declare(strict_types=1);

namespace Alxarafe\App\Domain\Topology\ValueObject;

use InvalidArgumentException;

/**
 * Value Object representing a zone type code (e.g. PICKING, BULK, RETURNS).
 */
final readonly class ZoneTypeCode
{
    public function __construct(private string $value)
    {
        if ($value === '') {
            throw new InvalidArgumentException('ZoneTypeCode cannot be empty.');
        }

        if (preg_match('/^[A-Z][A-Z0-9_]{1,19}$/', $value) !== 1) {
            throw new InvalidArgumentException(
                'ZoneTypeCode must be uppercase alphanumeric with underscores (2-20 chars). Got: ' . $value
            );
        }
    }

    public function value(): string
    {
        return $this->value;
    }

    public function equals(self $other): bool
    {
        return $this->value === $other->value;
    }

    public function __toString(): string
    {
        return $this->value;
    }
}

File: java/src/main/java/com/alxarafe/app/domain/topology/valueobject/ZoneTypeCode.java
package com.alxarafe.app.domain.topology.valueobject;

import java.util.Objects;
import java.util.regex.Pattern;

/**
 * Value Object representing a zone type code (e.g. PICKING, BULK, RETURNS).
 */
public record ZoneTypeCode(String value) {

    private static final Pattern FORMAT = Pattern.compile("^[A-Z][A-Z0-9_]{1,19}$");

    public ZoneTypeCode {
        Objects.requireNonNull(value, "ZoneTypeCode cannot be null.");
        if (!FORMAT.matcher(value).matches()) {
            throw new IllegalArgumentException(
                    "ZoneTypeCode must be uppercase alphanumeric with underscores (2-20 chars). Got: " + value);
        }
    }

    @Override
    public String toString() {
        return value;
    }
}
